x/rollapp/genesisbridge/ibc_module.go

```diff
@@ -107,42 +105,41 @@ func (w IBCModule) OnRecvPacket(
 		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "unmarshal genesis bridge data"))
 	}
 
-	// stateless validation of the genesis bridge data
-	if err := genesisBridgeData.ValidateBasic(); err != nil {
-		l.Error("Validate basic genesis bridge data.", "err", err)
-		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "validate basic genesis bridge data"))
+	v := validator{
+		rollapp:   genesisBridgeData,
+		hub:       ra.GenesisInfo,
+		channelID: ra.ChannelId,
+		rollappID: ra.RollappId,
 	}
 
-	// validate genesis info against the expected data set on the rollapp
-	err = w.ValidateGenesisBridge(ra, genesisBridgeData.GenesisInfo)
+	actionItems, err := v.validateAndGetActionItems()
 	if err != nil {
-		l.Error("Validate genesis info.", "err", err)
-		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "validate genesis info"))
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "validate and get actionable data"))
 	}
 
-	// register the denom metadata. the supplied denom is validated on validateBasic
-	raBaseDenom, err := w.registerDenomMetadata(ctx, ra.RollappId, ra.ChannelId, genesisBridgeData.NativeDenom)
-	if err != nil {
-		l.Error("Register denom metadata.", "err", err)
-		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "transfer genesis: register denom metadata"))
+	w.transferKeeper.SetDenomTrace(ctx, actionItems.trace)
+	if err := w.denomKeeper.CreateDenomMetadata(ctx, actionItems.bankMeta); err != nil {
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "create denom metadata"))
 	}
 
-	// validate and handle the genesis transfer
-	err = w.handleGenesisTransfer(ctx, *ra, packet, genesisBridgeData.GenesisTransfer)
-	if err != nil {
-		l.Error("Handle genesis transfer.", "err", err)
+	if err := w.transferKeeper.OnRecvPacket(ctx, packet, actionItems.fungiDatas[0]); err != nil {
 		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "handle genesis transfer"))
 	}
+	for _, data := range actionItems.fungiDatas {
+		if err := w.transferKeeper.OnRecvPacket(ctx, packet, data); err != nil {
+			return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "handle genesis transfer"))
+		}
+	}
```

Comment:
Here, we still have three points of failure, even after validation
1. Set denom trace in transfer keeper
2. Recv the packet(s) downstream
3. Create the metadata  in denom keeper

It would be hard to simulate these because they need the real channel and packet. But given that we do so much validation beforehand, it might be a good practical solution